### ISSUE TITLE
Add a meta annotation for marking a class/function as "package private"

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/linter#3606]
**Describe the rule you'd like to see implemented**
When writing a package, it is common to define a public class that should **not** be exported by the package.

package:meta comes with various utilities like `@protected` or `visibleForTesting`, but these aren't scoped to a package.
For example, a "protected" method could still be used outside of the defining package by subclassing its class.

Similarly, Dart offers a way for exports to only export a few specific types/variable via `export ... show/hide`. But this is error prone.
Using `show` could lead to the package forgetting to export a newly defined type. And using `hide` could lead to the package forgetting to hide a package-private class.

I believe a custom `@packageProtected` annotation that could be applied on anything could solve this problem.

**Examples**

I see such annotation have a few effects.

First, the annotation would trigger a warning if used outside of the defining package:

```dart
// package:foo/src/foo.dart
class Foo {
  final int a;
  @packageProtected
  final int b;
}
```

Then we'd have:

```dart
// package:bar/bar.dart
void main() {
  var foo = Foo();
  foo.a; // OK
  foo.b; // KO, b should only be used by `package:foo`
}

// package:foo/src/another.dart
void main() {
  var foo = Foo();
  foo.a; // OK
  foo.b; // OK since part of the same package
}
```

Secondly, I would expect `export` statements at the root of the `lib` folder to spot mistakes.

Consider:

```dart
// package:foo/src/foo.dart

@packageProtected
class  A { }

@packageProtected
var b;

var c;
```

Then for exports at the root of the `lib` folder, we'd have:

```dart
// package:foo/foo.dart
export 'src/foo.dart'; // KO, `A` and `b` should be hidden
export 'src/foo.dart' hide b; // KO, we properly have hidden `b` but forgot to hide `A`
export 'src/foo.dart' show A; // KO, A is packageProtected and should not be exported
export 'src/foo.dart' show c; // no problem
```

This would on the other hand not affect exports that are _not_ at the root of the lib folder:

```dart
// package:foo/src/not_public_file.dart
export 'src/foo.dart'; no problem here
```

Finally, I think it would also help to warn if `@packageProtected` types are part of the public API:

```dart 
@packageProtected
class A {}

void publicFn(A a) {} // KO, publicFn is not packageProtected but A is

void _private(A a) {} // no problem

@packageProtected
void protectedFn(A a) {} // no problem either
```

**Comment by srawlins** [comment]
Is `@internal` different from what you want?

**Comment by rrousselGit** [comment]
I've never heard about `@internal` before. I'll check it out, thanks!

**Comment by rrousselGit** [comment]
It does look similar to what I want. Thanks again!

I'll close this for now then.

**Comment by rrousselGit** [comment]
Damn this is awesome, that's exactly what I wanted. Thanks again!

**Comment by srawlins** [comment]
Haha no problem! A request from the Angular and Sass teams, I believe
